### ISSUE TITLE
Request: more examples in the Walt Eplorer

### ISSUE THREAD

**Issue by JobLeonard** [ballercat/walt#101]
# Feature Request
## Overview
[The on-line Walt explorer](https://ballercat.github.io/walt/) has more than one function: it shows the power of Walt and WebAssembly, can give inspiration for what to use it for, and works as a pseudo-tutorial or toolkit  with "fundamental" solutions to specific problems. Adding more examples would help in all of these matters.

The "C-style union using arrays"-trick I used in #99 is an snippet that could be an advanced example. 

# Impact
Large (imo). First impressions matter. Also, it might be useful to note here that *I have no prior experience in WASM* (although I did [program a game in z80 once](http://www.ticalc.org/archives/files/fileinfo/328/32816.html), but nevertheless managed to figure out how to port a simple C-function using the few examples in the playground right now.

It also doubles as a "cookbook" to get people up and running, and the more examples the better.

## Details
Walt is a mixture of "it's just JavaScript" and "it's just WebAssembly". If it is more accessible than the latter, it would help if people could learn Walt without prior knowledge of WebAssembly. 

Writing documentation and tutorials is one option (issue #73 partially covers this), but the Walt explorer allows us to take a written tutorial and give it much more power. A well-documented piece of code can essentially work as an interactive tutorial. In the words of Bret Victor: it's time to [stop drawing dead fish!](https://vimeo.com/64895205).

Here are some example galleries employed by other languages & frameworks:

Pixi.js  examples gallery - http://pixijs.io/examples/#/basics/basic.js

Phaser.io examples gallery - https://phaser.io/examples

"A Tour Of Go" interactive tutorial - https://tour.golang.org/welcome/1

Go Playground - https://play.golang.org/

Basically, currently the explorer is like Go's on-line playground (minus the linking bit). Adding a dash of *A Tour Of Go* would make things even better.

Potential topics to cover:

- a [radix sort implementation](https://run.perf.zone/view/20180308-Radix-Sorts-U32-I32-no-minification-warmed-up-sort-cache-1520481917994). Would cover:
  - how to pass an array of data
  - how to return an array of data
- an example function that does string manipulation (say, count number of spaces in a string). Would cover:
  - passing non-trivial arguments
  - returning non-trivial data

## Due Date
Eh.. whenever someone has an example that fits it could be added, I guess?

**Comment by ballercat** [comment]
This is fantastic. I'll give it a read through tomorrow. Thanks for the other issue as well 👍

**Comment by JobLeonard** [comment]
> This is fantastic

No, this whole project is fantastic! I get to combine my love of low-level bit-fiddling and the ease and accessability of the browser environment? Oh my...

I edited in another potential example: Duff's Device.

If you have any pointers on how to pass/return an array of data (of unknown size), and of a string, then I'd love to have a go at writing these examples myself - I can actually really use that radix sort, since I'm already using the JS versions as an optimisation somewhere.

**Comment by JobLeonard** [comment]
Also, I have recently written some custom scatter plot code that contains what is likely [the fastest bit of software-based sprite blitting code](https://github.com/linnarsson-lab/loom-viewer/blob/master/client/plotters/scatterplot.js#[L1706-L1890]) available in JavaScript. Because I can't be bothered to learn WebGL but love bit-fiddling.

```js
/**
 * @param {tObject} context
 * @param {number} width
 * @param {number} height
 * @param {Uint32Array} xySprites
 * @param {Uint32Array} spriteData32
 */
BlitSprites.prototype.littleEndian = function (context, width, height, xySprites, spriteData32) {
	let {
		canvasDataU32,
	} = this;
	const {
		xMask,
		yShift,
		spriteLength,
	} = this;


	for (let cIdx = 0, // canvas index
		sp = 0, // sprite pixel
		cp = 0, // canvas pixel
		sa = 0, // sprite alpha
		ca = 0, // canvas alpha
		j = 0; j < xySprites.length;) {
		for (let xy32 = xySprites[j++],
			x = xy32 & 0xFFFF,
			y = (xy32 >>> 16) & 0x7FFF,
			spriteOffset = xySprites[j++],
			c0 = x + y * width | 0, // top-left corner of sprite on target canvas
			i = 0; i < spriteLength; i++) {
			sp = spriteData32[i + spriteOffset];
			sa = (sp >>> 24);
			if (sa) {
				// index in canvas array
				cIdx = c0 + (i & xMask) + (i >>> yShift) * width;
				// canvas pixel
				cp = canvasDataU32[cIdx];
				// sa is a byte, so in the range [0,256).
				// We introduce a tiny bias towards sa (the newer pixel),
				// to replace /255 with /256, which can be written as
				// >>8, which is faster
				ca = 0x100 - ++sa;
				// A bit of bitmask-trickery lets us use:
				// - 4 instead of 6 multiplies
				// - 6 instead of 9 and-bitmasks
				// - 4 instead of 6 additions
				// - 3 instead of 1 logical shift
				canvasDataU32[cIdx] = (
					( // red + blue channel
						((sp & 0xFF00FF) * sa +
							(cp & 0xFF00FF) * ca) >>> 8
					) & 0xFF00FF
				) | ( // green channel
						(((sp >>> 8) & 0xFF) * sa +
						((cp >>> 8) & 0xFF) * ca) & 0xFF00
					) | 0xFF000000;
				/*
				ca = 0x100 - ++sa;
				// index in canvas array
				cIdx = c0 + (i & xMask) + (i >>> yShift) * width;
				// canvas pixel
				cp = canvasDataU32[cIdx];
				canvasDataU32[cIdx] = ((
					(
						(cp & 0xFF0000) * ca +
						(sp & 0xFF0000) * sa & 0xFF000000
					) + (
						(cp & 0xFF00) * ca +
						(sp & 0xFF00) * sa & 0xFF0000
					) + (
						(cp & 0xFF) * ca +
						(sp & 0xFF) * sa & 0xFF00
					)
				) >>> 8) + 0xFF000000;
				*/
			}
		}
	}
	// put imagedata back on the canvas
	context.putImageData(this.canvasData, 0, 0);
};
``` 

Would love to port it to Walt and see if it makes things even faster :D. If so I think that would make a cool example as well.

Again, all I really need is a way to quickly pass Uint32Arrays (or at least the data in it) between WASM and the environment.

**Comment by JobLeonard** [comment]
Ok, I just found out about the existence of WASMfiddle, through [this SO answer about strings](https://stackoverflow.com/a/47676844/3211791).

https://wasdk.github.io/WasmFiddle//?w590f

With your permission, I think I'll self-assign the string-example ;).

**Comment by JobLeonard** [comment]
> If you'd like to tackle additional examples feel free to go for it. 

If it's alright with you I'll use this issue in a "public notepad" for my thoughts on how to implement various examples then. As you might have noticed, I'm verbose (sorry), but I kind of work better when I try to write things out first before coding. And maybe it'll be an interesting read. Please jump in with suggestions from time to time :)

## Example: count occurrences of character in a string

> String literals are not implemented yet, so a traditional string example like above is not possible. 

True. But I was thinking of counting occurrences of a character in a string. This would not require a string literal just yet - the string is passed in from the outside, and the returned value is a number. It's the reverse of the situation explained in this SO answer: ["How can I return a JavaScript string from a WebAssembly function"](https://stackoverflow.com/a/41372484/3211791).

(Tangent: this example is purely that, an example. It is almost guaranteed to be slower than the JS-based solution - [`split().length-1`](https://jsperf.com/count-char-occurrences-switch) is beats every other option by a large margin, especially in Chrome. Also interesting is that Duff's Device is slower in Chrome, but faster in Firefox. However, this example works as a thought-exercise to figure out other string manipulations that *would* benefit from WASM. Plus, the process of working this problem out, I ended up opening the char and string issues, in case you're wondering where that came from)

Strings are UTF16-encoded in JavaScript, and individual  characters can be converted to numbers via [`charCodeAt`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/charCodeAt) (UTF16 codepoints, 16 bit number) or [`codePointAt(0)`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/codePointAt). The former is possibly faster (since we have less memory to go over), and the vast majority of characters will be 8 or 16 bits wide, but then we have the issue where `𠀬` (codepoint [U+2002C](https://en.wikibooks.org/wiki/Unicode/Character_reference/20000-20FFF)) gets mistaken for `,` (codepoint U+0002C)). How to handle that is discussed further on.

- the function is split into two parts: a JavaScript function (hereafter `countChar`), which prepares the string data, and calls the WASM part (hereafter `_countChar`)
- `_countChar` contains a `i32[]` array as buffer (with the default single memory page it can up to 16384 elements, more than enough for most cases)
- `countChar` uses that section of it as the backing array of a `Uint32Array` variable called `strBuffer` (native dwords are faster, and we don't have to worry about endianness this way)
- calling `countChar` will first iterate over the string, filling up `strBuffer` with at most 16384*2 characters at a time (packing two per element), until the end of the string is reached. This will be marked with zero in `strBuffer`, effectively becoming a zero-terminated string
- it then calls `_countChar`, which expects the UTF16 code of a character (for example, `,` is 44)
- `_countChar` counts the number of occurrences of that value in `strBuffer` until it encounters zero or runs out buffer, and returns that value.
- `countChar` adds this to the total count, and if it hasn't reached the end of the string yet continues with the next 256 characters.

### Dealing with UTF16 pairs

This is the example on [MDN's `charCodeAt` page](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/charCodeAt#Examples) on handling characters that take up more than 16 bits:

```js
  var code = str.charCodeAt(idx);
  var hi, low;
  
  // High surrogate (could change last hex to 0xDB7F
  // to treat high private surrogates 
  // as single characters)
  if (0xD800 <= code && code <= 0xDBFF) {
    hi = code;
    low = str.charCodeAt(idx + 1);
  }
```

So for any code point between 0xD800-0xDBFF, we know the next charcode is part of a pair. The easiest place to deal with this in the JavaScript wrapper: we iterate over the string and skip any code pair.

```js
// remember that strBuffer is a Uint32Array
let i = 0, j = 0;
while(i < str.length && j < strBuffer.length){
  let char = str.charCodeAt(i++) | 0;
  if (0xD800 <= char && char <= 0xDBFF) i++; // skip next char
  else strBuffer[j] = code << 16; // write char to buffer
  char = str.charCodeAt(i++) | 0; // if we went out of range, this forces `undefined` to 0
  if (0xD800 <= char && char <= 0xDBFF) i++; // skip next char
  else strBuffer[j++] = code; // write char to buffer
}
// ensure that the string is zero terminated
if (j < strBuffer.length) strBuffer[j] = 0;
```

.. of course, we need to write a special function for counting 32-bit characters, but how to handle that is probably fairly self-explanatory after this.
